<div class="paragraph">
<p>If a pointer to an object is used as an exception, the code that will catch the exception may or may not have to delete the pointed-to object. This is even more complex in the exception case than in classical manual memory management, because of the distance between the <code>throw</code> statements and the matching <code>catch</code>.</p>
</div>
<div class="paragraph">
<p>Throwing by value is just simpler and less error prone.</p>
</div>
<div class="sect1">
<h2 id="_noncompliant_code_example">Noncompliant Code Example</h2>
<div class="sectionbody">
<div class="listingblock">
<div class="content">
<pre>class E { /* Implementation */};
E globalException;

void fn ( int i )
{
  // In this situation, the developer writing the 'catch' has no way to know if the object pointed-to by
  // the exception should be deleted or not...
  if ( i &gt; 10 ) {
    throw ( &amp;globalException); // Noncompliant, the catch is supposed not to delete the pointer
  }
  else {
    throw (new E ); // Noncompliant, the catch is supposed to delete the pointer
  }
}</pre>
</div>
</div>
</div>
</div>
<div class="sect1">
<h2 id="_compliant_solution">Compliant Solution</h2>
<div class="sectionbody">
<div class="listingblock">
<div class="content">
<pre>class E { /* Implementation */};
E globalException;

void fn ( int i )
{
  if ( i &gt; 10 ) {
    throw ( globalException); // Throws a copy of the global variable
  }
  else {
    throw (E{} ); // Throws a new object
  }
}</pre>
</div>
</div>
</div>
</div>
<div class="sect1">
<h2 id="_see">See</h2>
<div class="sectionbody">
<div class="ulist">
<ul>
<li>
<p>MISRA C&#43;&#43; 2008, 15-0-2 - An exception object should not have pointer type.</p>
</li>
</ul>
</div>
<hr/>
</div>
</div>
<div class="sect1">
<h2 id="_implementation_specification">Implementation Specification</h2>
<div class="sectionbody">
<div class="paragraph">
<p>(visible only on this page)</p>
</div>
<div class="sect2">
<h3 id="_message">Message</h3>
<div class="paragraph">
<p>Throw the exception by value.</p>
</div>
<hr/>
</div>
</div>
</div>
<div class="sect1">
<h2 id="_comments_and_links">Comments And Links</h2>
<div class="sectionbody">
<div class="paragraph">
<p>(visible only on this page)</p>
</div>
<div class="sect2">
<h3 id="_on_14_oct_2014_175153_ann_campbell_wrote">on 14 Oct 2014, 17:51:53 Ann Campbell wrote:</h3>
<div class="paragraph">
<p>\[~samuel.mercier] please correct the following</p>
</div>
<div class="ulist">
<ul>
<li>
<p>Add a See section to the description listing the appropriate MISRA number</p>
</li>
<li>
<p>Fill in Applicability.</p>
</li>
<li>
<p>Make sure the appropriate MISRA C and MISRA C&#43;&#43; fields on the references tab are filled in</p>
</li>
</ul>
</div>
</div>
<div class="sect2">
<h3 id="_on_14_oct_2014_183418_ann_campbell_wrote">on 14 Oct 2014, 18:34:18 Ann Campbell wrote:</h3>
<div class="paragraph">
<p>Also [~samuel.mercier] I don&#8217;t think Maintainability is appropriate here. The risks are</p>
</div>
<div class="ulist">
<ul>
<li>
<p>both caller and callee destroy the object - undefined behavior (likely runtime crash)</p>
</li>
<li>
<p>neither caller nor callee destroys the object - memory leak
Since the worse outcome is a crash, I would associate this to Reliability</p>
</li>
</ul>
</div>
</div>
<div class="sect2">
<h3 id="_on_14_oct_2014_201526_freddy_mallet_wrote">on 14 Oct 2014, 20:15:26 Freddy Mallet wrote:</h3>
<div class="paragraph">
<p>And if this is associated to the Reliability category, then the tag 'pitfall' should be replaced by 'bug'</p>
</div>
</div>
<div class="sect2">
<h3 id="_on_17_oct_2014_091245_samuel_mercier_wrote">on 17 Oct 2014, 09:12:45 Samuel Mercier wrote:</h3>
<div class="paragraph">
<p>\[~ann.campbell.2] caller cannot destroys the object, unless it also catches the exception it has thrown. But I agree about the memory leak, so I updated to Reliablity / Exception handling</p>
</div>
</div>
<div class="sect2">
<h3 id="_on_20_apr_2015_100042_evgeny_mandrikov_wrote">on 20 Apr 2015, 10:00:42 Evgeny Mandrikov wrote:</h3>
<div class="paragraph">
<p>\[~ann.campbell.2] this rule has tag "bug" and severity "critical", but not enabled by default. Is it on purpose?</p>
</div>
</div>
<div class="sect2">
<h3 id="_on_20_apr_2015_100905_ann_campbell_wrote">on 20 Apr 2015, 10:09:05 Ann Campbell wrote:</h3>
<div class="paragraph">
<p>It was on purpose [~evgeny.mandrikov], since this strikes me as potentially an issue of house style. If you think it is not, feel free to set it on by default.</p>
</div>
</div>
<div class="sect2">
<h3 id="_on_11_may_2015_202444_evgeny_mandrikov_wrote">on 11 May 2015, 20:24:44 Evgeny Mandrikov wrote:</h3>
<div class="paragraph">
<p>\[~ann.campbell.2] I don&#8217;t see any valid reason for catching/throwing an exception by pointer, , while opposite is highly recommended, so don&#8217;t see any reason to not activate this rule by default.</p>
</div>
</div>
</div>
</div>